File: wpilibj/src/main/java/com/studica/frc/Cobra.java
// Copyright (c) 2026 WPILib contributors.

package com.studica.frc;

import edu.wpi.first.hal.JNIWrapper;

/** Studica Cobra four-channel reflectance array vendor API. */
public final class Cobra extends JNIWrapper implements AutoCloseable {
  public static final int CHANNEL_COUNT = 4;
  public static final double DEFAULT_REFERENCE_VOLTAGE = 5.0;

  private long handle;
  private int lastStatus;

  /** Creates a Cobra using the driver's reference-voltage semantics. */
  public Cobra() {
    this(DEFAULT_REFERENCE_VOLTAGE);
  }

  public Cobra(double referenceVoltage) {
    long[] result = create(referenceVoltage);
    if (result == null || result.length < 2) {
      handle = 0;
      lastStatus = -5;
    } else {
      handle = result[0];
      lastStatus = (int) result[1];
    }
  }

  public synchronized int getRaw(int channel) {
    if (channel < 0 || channel >= CHANNEL_COUNT) {
      lastStatus = -22;
      return -1;
    }
    return handle == 0 ? -1 : getRaw(handle, channel);
  }

  public synchronized double getVoltage(int channel) {
    if (channel < 0 || channel >= CHANNEL_COUNT) {
      lastStatus = -22;
      return 0.0;
    }
    return handle == 0 ? 0.0 : getVoltage(handle, channel);
  }

  public synchronized int getChannelCount() {
    return handle == 0 ? 0 : getChannelCount(handle);
  }

  public synchronized double getReferenceVoltage() {
    return handle == 0 ? 0.0 : getReferenceVoltage(handle);
  }

  public synchronized boolean isAvailable() {
    return handle != 0 && isAvailable(handle);
  }

  public synchronized int getLastStatus() {
    return lastStatus;
  }

  @Override
  public synchronized void close() {
    if (handle != 0) {
      destroy(handle);
      handle = 0;
    }
  }

  private static native long[] create(double referenceVoltage);

  private static native void destroy(long handle);

  private static native int getRaw(long handle, int channel);

  private static native double getVoltage(long handle, int channel);

  private static native int getChannelCount(long handle);

  private static native double getReferenceVoltage(long handle);

  private static native boolean isAvailable(long handle);
}

File: hal/src/main/native/cpp/Cobra.cpp
// Copyright (c) 2026 WPILib contributors.

#include "studica/Cobra.h"

#include <utility>

namespace studica {

Cobra::Cobra(double referenceVoltage) noexcept {
  m_lastStatus = StudicaCobra_Create(referenceVoltage, &m_handle);
}

Cobra::~Cobra() { StudicaCobra_Destroy(m_handle); }

Cobra::Cobra(Cobra&& other) noexcept
    : m_handle{std::exchange(other.m_handle, 0)},
      m_lastStatus{std::exchange(other.m_lastStatus,
                                 STUDICA_COBRA_NOT_INITIALIZED)} {}

Cobra& Cobra::operator=(Cobra&& other) noexcept {
  if (this == &other) return *this;
  StudicaCobra_Destroy(m_handle);
  m_handle = std::exchange(other.m_handle, 0);
  m_lastStatus = std::exchange(other.m_lastStatus,
                               STUDICA_COBRA_NOT_INITIALIZED);
  return *this;
}

int32_t Cobra::GetRaw(uint8_t channel) const noexcept {
  int32_t value = 0;
  m_lastStatus = StudicaCobra_GetRaw(m_handle, channel, &value);
  return m_lastStatus == STUDICA_COBRA_OK ? value : -1;
}

double Cobra::GetVoltage(uint8_t channel) const noexcept {
  double value = 0.0;
  m_lastStatus = StudicaCobra_GetVoltage(m_handle, channel, &value);
  return m_lastStatus == STUDICA_COBRA_OK ? value : 0.0;
}

uint8_t Cobra::GetChannelCount() const noexcept {
  uint8_t value = 0;
  m_lastStatus = StudicaCobra_GetChannelCount(m_handle, &value);
  return m_lastStatus == STUDICA_COBRA_OK ? value : 0;
}

double Cobra::GetReferenceVoltage() const noexcept {
  double value = 0.0;
  m_lastStatus = StudicaCobra_GetReferenceVoltage(m_handle, &value);
  return m_lastStatus == STUDICA_COBRA_OK ? value : 0.0;
}

bool Cobra::IsAvailable() const noexcept {
  uint8_t value = 0;
  m_lastStatus = StudicaCobra_IsAvailable(m_handle, &value);
  return m_lastStatus == STUDICA_COBRA_OK && value != 0;
}

int32_t Cobra::GetLastStatus() const noexcept { return m_lastStatus; }

}  // namespace studica
